src/repr/src/adt/interval.rs

```diff
@@ -551,6 +567,98 @@ impl Interval {
 
         Some(res)
     }
+
+    /// Adjust interval so 'days' contains less than 30 days, adding the excess to 'months'.
```

Comment:
Most of this logic comes directly from Postgres: https://github.com/postgres/postgres/blob/73c61a50a1555007001d29844dcdb10b4f982a73/src/backend/utils/adt/timestamp.c#L2695-L2825

And an approved patch that removes some overflow issues from the postgres code: https://commitfest.postgresql.org/37/3540/